File: table.rb
require 'gtk2'
require 'signalregistry'

module Cpg
	class Table < Gtk::Table
		type_register
	
		include SignalRegistry
	
		attr_accessor :expand

		EXPAND_RIGHT = 1
		EXPAND_DOWN = 2
	
		class Placeholder < Gtk::DrawingArea
			type_register
		
			def initialize(*args)
				super
			
				show
			end
		end
	
		def initialize(rows = 1, cols = 1, homogeneous = false)
			super({'n-rows' => rows, 'n-columns' => cols, 'homogeneous' => homogeneous})
		
			@expand = EXPAND_RIGHT
			@signals = {}
		
			add_events(Gdk::Event::KEY_PRESS_MASK | Gdk::Event::BUTTON_PRESS_MASK)
			set_can_focus(true)

			Gtk::Drag.dest_set(self, 0, [['TableItem', Gtk::Drag::TARGET_SAME_APP, 1]], Gdk::DragContext::ACTION_MOVE)
		end
	
		def child_position(child, left, top)
			child_set_property(child, 'left_attach', left)
			child_set_property(child, 'right_attach', left + 1)
			child_set_property(child, 'top_attach', top)
			child_set_property(child, 'bottom_attach', top + 1)
		end
	
		def move_to(from, to)
			['left', 'right', 'top', 'bottom'].each do |name|
				child_set_property(from, "#{name}_attach", child_get_property(to, "#{name}_attach"))
			end
		end
	
		def do_update_dragging(x, y)
			# we have the child in @dragging which was dragged from @dragpos, we
			# need to determine the child position under the cursor at x, y
			# then swap accordingly
		
			mtx = Array.new(self.n_columns)
			mtx.collect! { |idx| Array.new(self.n_rows) }
		
			columns = Array.new(self.n_columns)
			rows = Array.new(self.n_rows)
		
			children.each do |child|
				a = child.allocation
				left = child_get_property(child, 'left_attach')
				top = child_get_property(child, 'top_attach')
			
				columns[left] = [a.x, a.x + a.width]
				rows[top] = [a.y, a.y + a.height]
			
				mtx[left][top] = child
			end

			col = nil
			columns.each_with_index do |ptr, idx|
				if ptr[0] < x && ptr[1] > x
					col = idx 
					break
				end
			end
		
			row = nil
			rows.each_with_index do |ptr, idx|
				if ptr[0] < y && ptr[1] > y
					row = idx
					break
				end
			end
		
			return unless (row && col)
		
			found = mtx[col][row]
			return if found == @dragging
		
			if @swapped
				# first make sure to move current swapped back to position drag
				# is now at
				move_to(@swapped, @dragging)
			end
		
			if @swapped && @swapped == found
				found = @dragging
				@swapped = nil
			else
				# now move @dragging to new position
				child_position(@dragging, col, row)
		
				# set swapped to newly found
				@swapped = found
			end
		
			child_position(found, @dragpos[0], @dragpos[1]) if found
		end
	
		def find_empty
			all = []
		
			(0..self.n_columns - 1).each do |x|
				(0..self.n_rows - 1).each do |y|
					all << [x, y]
				end
			end
		
			if @expand == EXPAND_DOWN
				all.sort! { |a, b| a[1] == b[1] ? a[0] <=> b[0] : a[1] <=> b[1] }
			end

			children.each do |child|
				left = child_get_property(child, 'left_attach')
				top = child_get_property(child, 'top_attach')
			
				all.delete_if { |x| x == [left, top] }
			end	
		
			return all[0] unless all.empty?
		
			# otherwise we need to resize
			if @expand == EXPAND_RIGHT
				resize(self.n_rows, self.n_columns + 1)
				return [self.n_columns - 1, 0]
			else
				resize(self.n_rows + 1, self.n_columns)
				return [0, self.n_rows - 1]
			end
		end
	
		def <<(child)
			add(child)
		end
	
		def add(child)
			left, top = find_empty
		
			if (child.flags & Gtk::Widget::NO_WINDOW)
				ev = Gtk::EventBox.new
				child.show
			
				child.signal_connect('destroy') { |widget| ev.destroy }
				ev << child
				child = ev
			end
		
			attach(child, left, left + 1, top, top + 1, Gtk::EXPAND | Gtk::FILL, Gtk::EXPAND | Gtk::FILL, 0, 0)
		
			set_drag_source(child)
			child.show
		end

		def set_drag_source(child)
			Gtk::Drag.source_set(child, Gdk::Window::ModifierType::BUTTON1_MASK, [['TableItem', Gtk::Drag::TARGET_SAME_APP, 1]], Gdk::DragContext::ACTION_MOVE)

			connect(child, 'drag-begin') { |c, ctx| do_drag_begin(c, ctx) }
			connect(child, 'drag-end') { |c, ctx| do_drag_end(c, ctx) }
		end
	
		def do_drag_begin(child, context)
			# create drag icon
			alloc = child.allocation
			pix = Gdk::Pixbuf.from_drawable(nil, child.window, 0, 0, alloc.width, alloc.height)

			Gtk::Drag.set_icon(context, pix, alloc.width / 2, alloc.height / 2)
			@dragging = child
			@swapped = nil
		
			@dragpos = [child_get_property(child, 'left-attach'),
						child_get_property(child, 'top-attach')]
		end
	
		def do_drag_end(child, context)
			@dragging = nil
			
			compact
		end
		
		def empty_row?(row)
			!children.find { |child| !child.is_a?(Placeholder) && child_get_property(child, 'top-attach') == row }
		end
		
		def empty_column?(col)
			!children.find { |child| !child.is_a?(Placeholder) && child_get_property(child, 'left-attach') == col }
		end
		
		def find_empty_row
			return nil if self.n_rows <= 1
			(0..self.n_rows - 1).find { |x| empty_row?(x) }
		end
		
		def find_empty_column
			return nil if self.n_columns <= 1
			(0..self.n_columns - 1).find { |x| empty_column?(x) }
		end
		
		def remove_column(col)
			# move all children with left-attach > col to -1
			children.each do |child| 
				left = child_get_property(child, 'left-attach')
				
				if left == col && child.is_a?(Placeholder)
					child.destroy
				else
					child_set_property(child, 'left-attach', left - 1) if left && left > col
				end
			end
			
			resize(self.n_rows, self.n_columns - 1)
			
			queue_draw
		end
		
		def remove_row(row)
			# move all children with top-attach > row to -1
			children.each do |child| 
				top = child_get_property(child, 'top-attach')
				
				if top == row && child.is_a?(Placeholder)
					child.destroy
				else
					child_set_property(child, 'top-attach', top - 1) if top && top > row
				end
			end
			
			resize(self.n_rows - 1, self.n_columns)
			
			queue_draw
		end

		def compact
			return if destroyed?

			# remove any empty rows or columns
			while ((r = find_empty_row))
				remove_row(r)
			end
			
			while ((c = find_empty_column))
				remove_column(c)
			end
		end
		
		def signal_do_destroy
			@destroying = true
			super
		end
	
		def signal_do_remove(w)
			super
			
			return if @destroying
			
			(@signals[w] || []).each { |s| w.signal_handler_disconnect(s) }
			@signals.delete(w)
			
			compact unless w.is_a?(Placeholder)
		end
	
		def find_empty_row_col(last, what)
			placeholder = nil
			ret = children.find do |child| 
				if child_get_property(child, "#{what}-attach") == last
					r = !child.is_a?(Placeholder)
					placeholder = child unless r
				
					r
				else
					false
				end
			end
		
			return [ret, placeholder]
		end
	
		def remove_empty_row
			last = self.n_rows - 1
			ret, placeholder = find_empty_row_col(last, 'top')
		
			return false if ret
		
			placeholder.destroy if placeholder
			remove_row(last)

			true
		end
	
		def remove_empty_column
			last = self.n_columns - 1
			ret, placeholder = find_empty_row_col(last, 'left')
		
			return false if ret
		
			placeholder.destroy if placeholder
			remove_column(last)

			true
		end
	
		def signal_do_key_press_event(ev)
			return false unless ev.state.control_mask?

			case ev.keyval
				when Gdk::Keyval::GDK_Down
					return remove_empty_row
				when Gdk::Keyval::GDK_Up
					resize(self.n_rows + 1, self.n_columns)
					attach(Placeholder.new, 0, 1, self.n_rows - 1, self.n_rows)
					queue_draw
					return true
				when Gdk::Keyval::GDK_Left
					resize(self.n_rows, self.n_columns + 1)
					attach(Placeholder.new, self.n_columns - 1, self.n_columns, 0, 1)
					queue_draw
					return true
				when Gdk::Keyval::GDK_Right
					return remove_empty_column
			end
		
			false
		end
	
		def signal_do_drag_motion(ctx, x, y, time)
			super
		
			if !@dragging
				ctx.drag_status(0, time)
				false
			else
				do_update_dragging(x, y)
				ctx.drag_status(Gdk::DragContext::ACTION_MOVE, time)
				true
			end
		end
		
		def signal_do_button_press_event(ev)
			grab_focus
			false
		end
		
		def signal_do_drag_drop(ctx, x, y, time)
			return false unless @dragging
			
			Gtk::Drag.finish(ctx, true, false, time)
		end
	end
end

File: Table.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace Cpg.Studio.GtkGui
{
	public class Table : Gtk.Table
	{
		class Placeholder : Gtk.DrawingArea
		{
			public Placeholder() : base()
			{
				Show();
			}
		}
		
		enum Expand
		{
			Right,
			Down
		}
		
		private Expand d_expand;
		private bool d_compacting;
		private Gtk.Widget d_dragging;
		private Gtk.Widget d_swapped;
		private Point d_dragPos;
		
		public Table(uint rows, uint cols, bool homogeneous) : base(rows, cols, homogeneous)
		{
			d_expand = Expand.Right;
			
			AddEvents((int)(Gdk.EventMask.KeyPressMask | Gdk.EventMask.ButtonPressMask));
			CanFocus = true;
			
			Gtk.Drag.DestSet(this, 0, new Gtk.TargetEntry[] {new Gtk.TargetEntry("TableItem", Gtk.TargetFlags.App, 1)}, Gdk.DragAction.Move);
		}
		
		public Table() : this(1, 1, false)
		{
		}
		
		private Gtk.Widget RealChild(Gtk.Widget child)
		{
			while (child != null && child.Parent != this)
				child = child.Parent;
			
			return child;
		}
		
		private void PositionChild(Gtk.Widget child, uint left, uint top)
		{
			child = RealChild(child);

			ChildSetProperty(child, "left-attach", new GLib.Value(left));
			ChildSetProperty(child, "right-attach", new GLib.Value(left + 1));
			ChildSetProperty(child, "top-attach", new GLib.Value(top));
			ChildSetProperty(child, "bottom-attach", new GLib.Value(top + 1));
		}
		
		private void MoveTo(Gtk.Widget from, Gtk.Widget to)
		{
			foreach (string attr in new string[] {"left", "right", "top", "bottom"})
				ChildSetProperty(from, attr + "-attach", ChildGetProperty(to, attr + "-attach"));
		}
		
		private void DoUpdateDragging(int x, int y)
		{
			/* We have the child in d_dragging which was dragged from d_dragPos. We need to
			 * determine the child position under the cursor at x, y and swap accordingly
			 */
			Gtk.Widget[,] mtx = new Gtk.Widget[NColumns, NRows];
			
			object[] columns = new object[NColumns];
			object[] rows = new object[NRows];
			
			foreach (Gtk.Widget child in Children)
			{
				Gdk.Rectangle alloc = child.Allocation;
				
				uint left = (uint)ChildGetProperty(child, "left-attach").Val;
				uint top = (uint)ChildGetProperty(child, "top-attach").Val;
				
				columns[left] = new Point(alloc.X, alloc.X + alloc.Width);
				rows[top] = new Point(alloc.Y, alloc.Y + alloc.Height);
				
				mtx[left, top] = child;
			}
			
			int col = -1;
			
			for (int i = 0; i < NColumns; ++i)
			{
				if (columns[i] == null)
					continue;
				
				Point pt = (Point)columns[i];
				
				if (pt.X < x && pt.Y > x)
				{
					col = i;
					break;
				}
			}
			
			int row = -1;
			
			for (int i = 0; i < NRows; ++i)
			{
				if (rows[i] == null)
					continue;
				
				Point pt = (Point)rows[i];
				
				if (pt.X < y && pt.Y > y)
				{
					row = i;
					break;
				}
			}
			
			if (row == -1 || col == -1)
				return;
			
			Gtk.Widget found = mtx[col, row];
			
			if (found == d_dragging)
				return;
			
			if (d_swapped != null)
				MoveTo(d_swapped, d_dragging);
			
			if (d_swapped != null && d_swapped == found)
			{
				found = d_dragging;
				d_swapped = null;
			}
			else
			{
				PositionChild(d_dragging, (uint)col, (uint)row);
				d_swapped = found;
			}
			
			if (found != null)
				PositionChild(found, (uint)d_dragPos.X, (uint)d_dragPos.Y);
		}
		
		private Point FindEmpty()
		{
			List<Point> all = new List<Point>();
			
			for (int x = 0; x < NColumns; ++x)
			{
				for (int y = 0; y < NRows; ++y)
					all.Add(new Point(x, y));
			}
			
			if (d_expand == Expand.Down)
			{
				all.Sort(delegate (Point first, Point second) {
					return first.Y == second.Y ? first.X.CompareTo(second.X) : first.Y.CompareTo(second.Y);
				});
			}
			
			foreach (Gtk.Widget child in Children)
			{
				if (child is Placeholder)
					continue;
			
				Point pos = ChildPosition(child);
				
				all.RemoveAll(delegate (Point pt) {
					return pt == pos;
				});
			}
			
			if (all.Count > 0)
				return all[0];
			
			/* Resize is imminent */
			if (d_expand == Expand.Right)
			{
				Resize(NRows, NColumns + 1);
				return new Point((int)NColumns - 1, 0);
			}
			else
			{
				Resize(NRows + 1, NColumns);
				return new Point(0, (int)NRows - 1);
			}
		}
		
		private Point ChildPosition(Gtk.Widget child)
		{
			child = RealChild(child);
			
			if (child == null)
				return new Point();
			
			int left = (int)ChildGetProperty(child, "left-attach").Val;
			int top = (int)ChildGetProperty(child, "top-attach").Val;
			
			return new Point(left, top);
		}
		
		public new void Add(Gtk.Widget child)
		{
			Point pos = FindEmpty();

			if (child.IsNoWindow)
			{
				Gtk.EventBox ev = new Gtk.EventBox();
				child.Show();
				
				child.Destroyed += delegate (object source, EventArgs args) { ev.Destroy(); };
				ev.Add(child);
				
				child = ev;
			}
			
			Attach(child, (uint)pos.X, (uint)pos.X + 1, (uint)pos.Y, (uint)pos.Y + 1, Gtk.AttachOptions.Expand | Gtk.AttachOptions.Fill, Gtk.AttachOptions.Expand | Gtk.AttachOptions.Fill, 0, 0);
			
			SetDragSource(child);
			child.Show();
		}
		
		private void SetDragSource(Gtk.Widget child)
		{
			Gtk.Drag.SourceSet(child, Gdk.ModifierType.Button1Mask, new Gtk.TargetEntry[] {new Gtk.TargetEntry("TableItem", Gtk.TargetFlags.App, 1)}, Gdk.DragAction.Move);
			
			child.DragBegin += delegate (object source, Gtk.DragBeginArgs args) { DoDragBegin(child, args.Context); };
			child.DragEnd += delegate (object source, Gtk.DragEndArgs args) { DoDragEnd(child, args.Context); };
		}
		
		private void DoDragBegin(Gtk.Widget child, Gdk.DragContext context)
		{
			Gdk.Rectangle alloc = child.Allocation;
			Gdk.Pixbuf icon;
			
			System.Reflection.MethodInfo info = child.GetType().GetMethod("CreateDragIcon");
			
			if (info != null)
				icon = info.Invoke(child, new object[] {}) as Gdk.Pixbuf;
			else
				icon = Gdk.Pixbuf.FromDrawable(child.GdkWindow, child.GdkWindow.Colormap, 0, 0, 0, 0, alloc.Width, alloc.Height);
					
			Gtk.Drag.SetIconPixbuf(context, icon, icon.Width / 2, icon.Height / 2);
			
			d_dragging = child;
			d_swapped = null;
			
			d_dragPos = ChildPosition(child);
		}
		
		private bool EmptyRow(uint idx)
		{
			return (new List<Gtk.Widget>(Children)).Exists(delegate (Gtk.Widget child) {
				return !(child is Placeholder) && (uint)ChildGetProperty(child, "top-attach") == idx;
			});
		}
		
		private bool EmptyColumn(uint idx)
		{
			return (new List<Gtk.Widget>(Children)).Exists(delegate (Gtk.Widget child) {
				return !(child is Placeholder) && (uint)ChildGetProperty(child, "left-attach") == idx;
			});
		}
		
		private bool FindEmptyRow(ref uint idx)
		{
			if (NRows <= 1)
				return false;
			
			for (idx = 0; idx < NRows; ++idx)
			{
				if (EmptyRow(idx))
					return true;
			}
			
			return false;
		}
		
		private bool FindEmptyColumn(ref uint idx)
		{
			if (NColumns <= 1)
				return false;
			
			for (idx = 0; idx < NColumns; ++idx)
			{
				if (EmptyColumn(idx))
					return true;
			}
			
			return false;							
		}

		private void RemoveRow(uint idx)
		{
			foreach (Gtk.Widget child in Children)
			{
				Point pos = ChildPosition(child);
				
				if (pos.Y == idx && child is Placeholder)
				{
					child.Destroy();
				}
				else if (pos.Y >= idx)
				{
					ChildSetProperty(child, "top-attach", new GLib.Value(idx - 1));
					ChildSetProperty(child, "bottom-attach", new GLib.Value(idx));
				}
			}
			
			Resize(NRows - 1, NColumns);
			QueueDraw();
		}
		
		private void RemoveColumn(uint idx)
		{
			foreach (Gtk.Widget child in Children)
			{
				Point pos = ChildPosition(child);
				
				if (pos.X == idx && child is Placeholder)
				{
					child.Destroy();
				}
				else if (pos.X > idx)
				{
					ChildSetProperty(child, "left-attach", new GLib.Value(idx - 1));
					ChildSetProperty(child, "right-attach", new GLib.Value(idx));
				}
			}
			
			Resize(NRows, NColumns - 1);
			QueueDraw();
		}

		private void Compact()
		{
			d_compacting = true;
		
			uint idx = 0;
			while (FindEmptyRow(ref idx))
				RemoveRow(idx);
			
			while (FindEmptyColumn(ref idx))
				RemoveColumn(idx);
		
			d_compacting = false;
		}
		
		private void DoDragEnd(Gtk.Widget child, Gdk.DragContext context)
		{
			d_dragging = null;
			Compact();
		}
		
		protected override void OnRemoved(Gtk.Widget widget)
		{
			base.OnRemoved(widget);
			
			if (!d_compacting)
				Compact();
		}

		private void FindEmptyRowCol(uint last, string what, out Gtk.Widget child, out Placeholder placeholder)
		{
			placeholder = null;
			child = null;
			
			foreach (Gtk.Widget widget in Children)
			{
				if ((uint)ChildGetProperty(widget, what + "-attach").Val == last)
				{
					if (widget is Placeholder)
					{
						placeholder = widget as Placeholder;
					}
					else
					{
						child = widget;
						break;
					}
				}
			}
		}
		
		private bool RemoveEmptyRow()
		{
			uint last = NRows - 1;
			Gtk.Widget child;
			Placeholder placeholder;
			
			FindEmptyRowCol(last, "top", out child, out placeholder);
			
			if (child != null)
				return false;
		
			if (placeholder != null)
				placeholder.Destroy();
		
			RemoveRow(last);
			
			return true;
		}
		
		private bool RemoveEmptyColumn()
		{
			uint last = NColumns - 1;
			Gtk.Widget child;
			Placeholder placeholder;
			
			FindEmptyRowCol(last, "left", out child, out placeholder);
			
			if (child != null)
				return false;
		
			if (placeholder != null)
				placeholder.Destroy();
		
			RemoveColumn(last);
			
			return true;
		}
		
		protected override bool OnKeyPressEvent(Gdk.EventKey evnt)
		{
			if ((evnt.State & Gdk.ModifierType.ControlMask) == 0)
				return false;
			
			switch (evnt.Key)
			{
				case Gdk.Key.Down:
					return RemoveEmptyRow();
				case Gdk.Key.Up:
					Resize(NRows + 1, NColumns);
					Attach(new Placeholder(), 0, 1, NRows - 1, NRows);
					QueueDraw();
					
					return true;
				case Gdk.Key.Left:
					Resize(NRows, NColumns + 1);
					Attach(new Placeholder(), NColumns - 1, NColumns, 0, 1);
					QueueDraw();
					
					return true;
				case Gdk.Key.Right:
					return RemoveEmptyColumn();
			}
			
			return false;
		}
		
		protected override bool OnDragMotion(Gdk.DragContext context, int x, int y, uint time_)
		{
			base.OnDragMotion(context, x, y, time_);
			
			if (d_dragging == null)
			{
				Gdk.Drag.Status(context, 0, time_);
				return false;
			}
			else
			{
				DoUpdateDragging(x, y);
				Gdk.Drag.Status(context, Gdk.DragAction.Move, time_);
				return true;
			}
		}
		
		protected override bool OnButtonPressEvent(Gdk.EventButton evnt)
		{
			GrabFocus();
			return false;
		}

		protected override bool OnDragDrop(Gdk.DragContext context, int x, int y, uint time_)
		{
			if (d_dragging == null)
				return false;
			
			Gtk.Drag.Finish(context, true, false, time_);
			return true;
		}
	}
}
